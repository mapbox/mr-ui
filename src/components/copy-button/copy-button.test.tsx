import React from 'react';
import CopyButton from './copy-button';
import { act, render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('CopyButton', () => {

  describe('basic', () => {
    const props = {
      text: 'Copied by the default test case'
    };

    test('renders as expected', () => {
      const { baseElement } = render(<CopyButton {...props} />)
      expect(baseElement).toMatchSnapshot();
    });

    test('changes state on click', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await waitFor(() => {
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      });
    });
  });

  describe('all props', () => {
    const mockOnCopy = jest.fn();
    const props = {
      text: 'more copiable text',
      onCopy: mockOnCopy,
      className: 'px6 py6 btn btn--purple btn--stroke',
      focusTrapPaused: true,
      block: true,
      passthroughProps: {
        'data-testid': 'foo'
      }
    };

    test('renders as expected', () => {
      const { baseElement } = render(<CopyButton {...props} />)
      expect(baseElement).toMatchSnapshot();
    });

    test('changes state on click', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await waitFor(() => {
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      });
    });

    test('calls onCopy callback', async () => {
      render(<CopyButton {...props} />)

      await act(async () => {
        await userEvent.click(screen.getByTestId('copy-button'));
      });

      expect(mockOnCopy).toHaveBeenCalledTimes(1);
      expect(mockOnCopy).toHaveBeenCalledWith('more copiable text');
    });
  });
});
