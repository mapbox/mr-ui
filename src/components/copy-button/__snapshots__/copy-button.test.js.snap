// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton Defaults changes state on click 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={true}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="check"
    />
    <Popover
      accessibleTitle={null}
      alignment="center"
      allowPlacementAxisChange={true}
      clickOutsideCloses={true}
      coloring="light"
      containWithinViewport={true}
      escapeCloses={true}
      getAnchorElement={[Function]}
      hasPointer={true}
      hideWhenAnchorIsOffscreen={true}
      observeSize={true}
      padding="small"
      placement="top"
      receiveFocus={true}
      trapFocus={false}
      zIndex={1}
    >
      <div
        className="txt-s"
      >
        Copied!
      </div>
    </Popover>
  </button>
</Tooltip>
`;

exports[`CopyButton Defaults changes state on click 2`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="clipboard"
    />
  </button>
</Tooltip>
`;

exports[`CopyButton Defaults renders as expected 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="clipboard"
    />
  </button>
</Tooltip>
`;

exports[`CopyButton all props changes state on click 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={true}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="check"
    />
    <Popover
      accessibleTitle={null}
      alignment="center"
      allowPlacementAxisChange={true}
      clickOutsideCloses={true}
      coloring="light"
      containWithinViewport={true}
      escapeCloses={true}
      getAnchorElement={[Function]}
      hasPointer={true}
      hideWhenAnchorIsOffscreen={true}
      observeSize={true}
      padding="small"
      placement="top"
      receiveFocus={true}
      trapFocus={false}
      zIndex={1}
    >
      <div
        className="txt-s"
      >
        Copied!
      </div>
    </Popover>
  </button>
</Tooltip>
`;

exports[`CopyButton all props changes state on click 2`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="clipboard"
    />
  </button>
</Tooltip>
`;

exports[`CopyButton all props renders as expected 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    aria-label="Copy"
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      name="clipboard"
    />
  </button>
</Tooltip>
`;
